Sources/SwiftProtobuf/TextDecoder.swift

```diff
@@ -21,67 +21,449 @@ import Swift
+public struct TextDecoder: Decoder {
+    internal var scanner: TextScanner
+    private var fieldCount = 0
+    private var terminator: UInt8?
+    private var fieldNameMap: FieldNameMap?
+    private var messageType: Message.Type?
 
-    internal init(text: String, extensions: ExtensionSet? = nil) {
+    internal var complete: Bool {
+        mutating get {
+            return scanner.complete
+        }
+    }
+
+    internal init(messageType: Message.Type, text: String, extensions: ExtensionSet?) throws {
         scanner = TextScanner(text: text, extensions: extensions)
+        guard let nameProviding = (messageType as? ProtoNameProviding.Type) else {
+            throw TextDecodingError.missingFieldNames
+        }
+        fieldNameMap = nameProviding._protobuf_fieldNames
+        self.messageType = messageType
     }
 
-    internal init(scanner: TextScanner) {
+    internal init(messageType: Message.Type, scanner: TextScanner, terminator: UInt8?) throws {
         self.scanner = scanner
+        self.terminator = terminator
+        guard let nameProviding = (messageType as? ProtoNameProviding.Type) else {
+            throw TextDecodingError.missingFieldNames
+        }
+        fieldNameMap = nameProviding._protobuf_fieldNames
+        self.messageType = messageType
     }
 
-    internal mutating func decodeFullObject<M: Message>(message: inout M, terminator: UInt8?) throws {
-        guard let nameProviding = (M.self as? ProtoNameProviding.Type) else {
-            throw DecodingError.missingFieldNames
+
+    public mutating func handleConflictingOneOf() throws {
+        throw TextDecodingError.conflictingOneOf
+    }
+
+    public mutating func nextFieldNumber() throws -> Int? {
+        if let terminator = terminator {
+            if scanner.skipOptionalObjectEnd(terminator) {
+                return nil
+            }
         }
-        let names = nameProviding._protobuf_fieldNames
-        while true {
-            if let terminator = terminator {
-                if scanner.skipOptionalObjectEnd(terminator) {
+        if fieldCount > 0 {
+            scanner.skipOptionalSeparator()
+        }
+        if let key = try scanner.nextOptionalExtensionKey() {
+            // Extension key; look up in the extension registry
+            if let fieldNumber = scanner.extensions?.fieldNumberForProto(messageType: messageType!, protoFieldName: key) {
+                fieldCount += 1
+                return fieldNumber
+            } else {
+                throw TextDecodingError.unknownField
+            }
+        } else if let fieldNumber = try scanner.nextFieldNumber(names: fieldNameMap!) {
+            fieldCount += 1
+            return fieldNumber
+        } else if terminator == nil {
+            return nil
+        } else {
+            throw TextDecodingError.truncated
+        }
+
+    }
+
+    public mutating func decodeSingularFloatField(value: inout Float) throws {
+        try scanner.skipRequiredColon()
```

Comment:
For JSON, the colon is skipped when we read the key.  But Text format is less consistent about when a colon is required -- message-valued fields don't require them, other types do -- so the colon needs to be swallowed here when we parse the value.